src/constraint_system/arithmetic.rs

```diff
@@ -169,6 +167,72 @@ impl<E: PairingEngine, P: TEModelParameters<BaseField = E::Fr>>
         c
     }
 
+    /// This gates turns on all the selctor polynomials to give users,
+    /// in some situations, the ability to use the extra selector for
+    /// more variables into additions.
+    ///
+    /// This type of gate is usually used when we need to have
+    /// the largest amount of performance and the minimum circuit-size
+    /// possible. Since it allows the end-user to setup all of the selector
+    /// coefficients.
+    ///
+    /// Equation: `(a*b)*q_m + a*q_l + b*q_r + d*q_4 + q_c + PI + q_o * c = 0`.
+    /// `d` will be set to zero if not provided.
+    ///
+    /// ### Returns
+    /// `c`
+    pub fn big_arith_gate(
+        &mut self,
+        a: Variable,
+        b: Variable,
+        c: Variable,
+        d: Option<Variable>,
+        q_m: E::Fr,
+        q_l: E::Fr,
+        q_r: E::Fr,
+        q_o: E::Fr,
+        q_c: E::Fr,
+        q_4: E::Fr,
+        pi: Option<E::Fr>,
+    ) -> Variable {
+        // Check if advice wire has a value
+        let d = match d {
+            Some(var) => var,
+            None => self.zero_var,
+        };
+
+        self.w_l.push(a);
+        self.w_r.push(b);
+        self.w_o.push(c);
+        self.w_4.push(d);
+
+        // Add selector vectors
+        self.q_m.push(q_m);
+        self.q_o.push(q_o);
+        self.q_c.push(q_c);
+        self.q_4.push(q_4);
+        self.q_l.push(q_l);
+        self.q_r.push(q_r);
+        self.q_arith.push(E::Fr::one());
+
+        self.q_range.push(E::Fr::zero());
+        self.q_logic.push(E::Fr::zero());
+        self.q_fixed_group_add.push(E::Fr::zero());
+        self.q_variable_group_add.push(E::Fr::zero());
+
+        if let Some(pi) = pi {
+            assert!(
+                self.public_inputs_sparse_store.insert(self.n, pi).is_none(),"The invariant of already having a PI inserted for this position should never exist"
+            );
+        }
```

Comment:
This should not be an error, it should remain a panic. Since it is the composer internally tracking the gate counter, it's actually an implementation error if the gate PI gets messed up, not the user's fault, in which case it's not a recoverable error. You can make the code more succinct by just refactoring this into its own function, something like:

```rust
fn increment_gate_counter_with_pi(&mut self, pi: Option<F>) {
    if let Some(pi) = pi {
        assert!(...)
    }
    self.n += 1;
}
```

or just refactor the assert with something like ```fn insert_public_input(&mut self, pi: Option<F>);```

Reply:
Also, whenever this panics, it's an indication of a failure in _another function_ not the current one. It's actually a nontrivial issue if this panic ever gets called. There should be a comment in the panic that says to file a bug. If we had more type-level checks we could get rid of this failure path completely.

Reply:
hmmm ye. I was thinking when I said that that we should panic above. Since the error propagation will cause an error anyway. But then the API would become a nighmare with `?` everywhere and it's tricky. So yes, you're right, I prefer the panic too!

Reply:
It's not just a convenience thing, it's also semantically wrong to propagate the error here and not panic.